draft-ietf-taps-interface.md

```diff
@@ -1780,38 +1772,170 @@ that the other endpoint is done sending on a connection.
-that a Send at one end of a Connection maps to exactly one Receive on the
-other end. This is possible with Protocol Stacks that provide
-message boundary preservation, but is not the case over Protocol Stacks that
-provide a simple octet stream transport.
+While many protocols can be represented as Message Framers, for the 
+purposes of the Transport Services interface these are ways for applications
+or application frameworks to define their own Message parsing to be
+included within a Connection's Protocol Stack. As an example, TLS can
+serve the purpose of framing data over TCP, but is exposed as a protocol
+natively supported by the Transport Services interface.
 
-For preserving message boundaries over stream transports, this interface
-provides receiver-side de-framing. This facility is based on the observation
-that, since many of our current application protocols evolved over TCP, which
-does not provide message boundary preservation, and since many of these protocols
-require message boundaries to function, each application layer protocol has
-defined its own framing. A Deframer allows an application to push this
-de-framing down into the interface, in order to transform an octet stream into
-a sequence of Messages.
+Most Message Framer protocols fall into one of two categories:
+- Header-prefixed record protocols, such as a basic Type-Length-Value (TLV) structure
+- Delimeter-separated protocols, such as HTTP/1.1.
 
-Concretely, receiver-side de-framing allows a caller to provide the interface
-with a function that takes an octet stream, as provided by the underlying
-Protocol Stack, reads and returns a single Message of an appropriate type for
-the application and platform, and leaves the octet stream at the start of the
-next Message to deframe. It consists of a Deframer Object with a single Action,
-Deframe. Since the Deframer depends on the protocol used at the application
-layer, it is bound to the Preconnection during the pre-establishment phase:
+Note that while Message Framers add the most value when placed above
+a protocol that otherwise does not preserve message boundaries, they can
+also be used with datagram- or message-based protocols. In these cases,
+they add an additional transformation to further encode or encapsulate,
+and can potentially support packing multiple application-layer Messages
+into individual transport datagrams.
+
+## Defining Message Framers
+
+Applications can define a Message Framer by creating a object to represent a unique
+definition of a framing protocol. Event handlers for inbound and outbound data
+are registered on this object.
 
 ~~~
-Preconnection.DeframeWith(Deframer)
+framer := NewMessageFramer()
+framer.setEventHandlers()
+~~~
+
+## Adding Message Framers to Connections
 
-{messageData} := Deframer.Deframe(OctetStream)
+The Message Framer object can then be added to one or more Preconnections
+to run on top of transport protocols. Multiple Framers may be added.
```

Comment:
How do we do the ordering of framers here?
How can I slide in framers later, e.g., for STARTTLS?

Reply:
Yes, you should be able to dynamically add. We'd need new text for that.

Reply:
Added text for this.